modules/dnn/src/darknet/darknet_io.cpp

```diff
@@ -574,17 +647,47 @@ namespace cv {
                             batch_normalize);
 
                         current_channels = filters;
+                        current_width = (current_width - kernel_size + 2*padding) / stride + 1;
+                        current_height = (current_height - kernel_size + 2*padding) / stride + 1;
+
+                    }
+                    else if (layer_type == "connected")
+                    {
+                        int output = getParam<int>(layer_params, "output", 1);
+                        bool batch_normalize = getParam<int>(layer_params, "batch_normalize", 0) == 1;
+
+                        CV_Assert(output > 0);
+
+                        setParams.setFullyConnected(output, batch_normalize);
+                        current_channels = 0;
+                        current_width = 0;
+                        current_height = 0;
+                        current_total = output;
+
                     }
                     else if (layer_type == "maxpool")
                     {
                         int kernel_size = getParam<int>(layer_params, "size", 2);
                         int stride = getParam<int>(layer_params, "stride", 2);
-                        int pad = getParam<int>(layer_params, "pad", 0);
+                        int pad = getParam<int>(layer_params, "padding", -1);
```

Comment:
It seems to me it should work in the same way as this https://github.com/opencv/opencv/pull/16358

Reply:
@dkurt thanks for your review. 
Can you please check output height and width formulae in case of maxpool and convolution layer. They seem different:
For Maxpool:
https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/src/maxpool_layer.c#L30-L31

For Convolution:
https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/src/convolutional_layer.c#L66-L74

Reply:
I checked output shape of maxpool layer on (width=6, height=6, channels=3) input image if **Padding is used same as Convolution layer**
```
[net]
batch=1
channels=3
height=6
width=6

[maxpool]
size=3
stride=2
padding=0
```
Output using Darknet : `2 x 2 x 3` (height x width x channels) `VALID` padding.
Output using OpenCV : `3 x 3 x 3` (height x width x channels) `SAME` padding.

In Darknet while parsing maxpool, there is no parameter `pad`,  `padding` parameter is used for MANUAL padding that makes output shape
```
l.out_w = (w + padding - size)/stride + 1;
l.out_h = (h + padding - size)/stride + 1;
```

**Please note**: 2*Padding is not there, unlike convolution padding.
`SAME` padding is applied when `padding` is not mentioned in model.
 
**Explanation of changes made for MaxPool padding:**
Hence I explicitly added 

- `pad_mode = SAME` in case no mention of `padding` in model. 
- `pad_mode = VALID` If found `padding = 0`
- `pad = (padding + 1) / 2` in case padding is not 0.

Please correct me if I missed something.

Reply:
@ashishkrshrivastava, According to this, `pad` is not used by Darknet but `padding` is parsed:

https://github.com/pjreddie/darknet/blob/b13f67bfdd87434e141af532cdb5dc1b8369aa3b/src/parser.c#L487

I think that it's better to use `size - 1` (`kernel_size - 1) as a default padding size similarly to Darknet. So at least it's more intuitive than `0` and `-1`

Reply:
@dkurt, using `int pad = getParam<int>(layer_params, "padding", kernel_size -1 );` wont solve the problem. 
Still if I use it then, 
padding will be set to `maxpool_param.set<int>("pad", padding / 2);` and this will give wrong result in some cases like 

1. `kernel_size = 2` , padding = 1 (kernel_size - 1) which makes pad = 0 (padding / 2).
2. `padding = 1`, pad = 0 (padding / 2)

whenever padding will be odd, results will be wrong. we can only handle even padding this way.

Reply:
There is one more issue, `padding = 0` does not work same as `VALID` padding. The reason is ceilmode is set true by default:

https://github.com/opencv/opencv/blob/5d01821d9f4d57921dad82045cc9819cfabcfe15/modules/dnn/src/layers/pooling_layer.cpp#L136

which makes output:

https://github.com/opencv/opencv/blob/5d01821d9f4d57921dad82045cc9819cfabcfe15/modules/dnn/src/layers/pooling_layer.cpp#L1165-L1166

on the other hand, output for` VALID` padding is:

https://github.com/opencv/opencv/blob/426482e05b9db0e891539d01177dc5be120af370/modules/dnn/src/layers/layers_common.cpp#L203-L207

Hence, setting ceilmode = false by default and setting padding = kernel_size - 1 (default) will work only for odd kernel_size i.e. even padding.